src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs

```diff
@@ -1337,4 +1357,158 @@ internal bool CanAddValue(HttpHeaderParser parser)
+
+            internal Enumerator(IEnumerable<KeyValuePair<HeaderDescriptor, object>> headerStore)
+            {
+                _headerStore = headerStore.GetEnumerator();
+                Current = default;
+            }
+
+            public bool MoveNext()
+            {
+                if (!_headerStore.MoveNext())
+                {
+                    return false;
+                }
+
+                KeyValuePair<HeaderDescriptor, object> nextPair = _headerStore.Current;
+
+                Current = new KeyValuePair<string, HeaderStringValues>(nextPair.Key.Name, ToStringValues(nextPair.Key, nextPair.Value));
+                return true;
+
+                static HeaderStringValues ToStringValues(HeaderDescriptor descriptor, object value)
+                {
+                    if (value is HttpHeaders.HeaderStoreItemInfo info)
+                    {
+                        if (info.OriginalRawValue != null)
+                        {
+                            return info.OriginalRawValue is string rawString ? new HeaderStringValues(rawString) : new HeaderStringValues((List<string>)info.OriginalRawValue);
+                        }
+
+                        return new HeaderStringValues(HttpHeaders.GetValuesAsStrings(descriptor, info));
+                    }
+
+                    Debug.Assert(value is string);
+                    return new HeaderStringValues((string)value);
+                }
+            }
+
+            public KeyValuePair<string, HeaderStringValues> Current { get; private set; }
+
+            public void Dispose()
+            {
+                _headerStore.Dispose();
+            }
+
+            object? IEnumerator.Current => Current;
+
+            void IEnumerator.Reset()
+            {
+                _headerStore.Reset();
+                Current = default;
+            }
+        }
+    }
+
+    public readonly struct HeaderStringValues : IEnumerable<string>
+    {
+        private readonly IEnumerable<string>? _values;
+        private readonly string? _value;
+
+        public HeaderStringValues(string value)
+        {
+            _value = value;
+            _values = null;
+        }
+
+        public HeaderStringValues(IEnumerable<string> value)
+        {
+            _values = value;
+            _value = null;
+        }
+
+        public Enumerator GetEnumerator() => new Enumerator(_values, _value);
+
+        IEnumerator<string> IEnumerable<string>.GetEnumerator() => GetEnumerator();
+
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        public struct Enumerator : IEnumerator<string>
+        {
+            private readonly IEnumerator<string>? _values;
+            private readonly string? _value;
+            private bool _completed;
+
+            public Enumerator(IEnumerable<string>? values, string? value)
+            {
+                _values = values?.GetEnumerator();
```

Comment:
I don't think this should be public.

Also, seems like it would be better to have two constructors here, one for single value, one for multi-value. That way the arguments don't need to be nullable.